src/wayland/data_device/device.rs

```diff
@@ -0,0 +1,173 @@
+use std::{cell::RefCell, os::unix::prelude::RawFd};
+
+use slog::debug;
+use wayland_server::{
+    protocol::{
+        wl_data_device::{self, WlDataDevice},
+        wl_data_offer::WlDataOffer,
+        wl_data_source::WlDataSource,
+        wl_seat::WlSeat,
+        wl_surface::WlSurface,
+    },
+    DelegateDispatch, DelegateDispatchBase, Dispatch, Resource,
+};
+
+use crate::wayland::{
+    compositor,
+    data_device::seat_data::{SeatData, Selection},
+    seat::Seat,
+    Serial,
+};
+
+use super::{dnd_grab, DataDeviceHandler, DataDeviceState, DataOfferUserData};
+
+static DND_ICON_ROLE: &str = "dnd_icon";
+
+/// Events that are generated by interactions of the clients with the data device
+#[derive(Debug)]
+pub enum DataDeviceEvent<D> {
+    /// A client has set the selection
+    NewSelection(Option<WlDataSource>),
+    /// A client started a drag'n'drop as response to a user pointer action
+    DnDStarted {
+        /// The data source provided by the client
+        ///
+        /// If it is `None`, this means the DnD is restricted to surfaces of the
+        /// same client and the client will manage data transfer by itself.
+        source: Option<WlDataSource>,
+        /// The icon the client requested to be used to be associated with the cursor icon
+        /// during the drag'n'drop.
+        icon: Option<WlSurface>,
+        /// The seat on which the DnD operation was started
+        seat: Seat<D>,
+    },
+    /// The drag'n'drop action was finished by the user releasing the buttons
+    ///
+    /// At this point, any pointer icon should be removed.
+    ///
+    /// Note that this event will only be generated for client-initiated drag'n'drop session.
+    DnDDropped {
+        /// The seat on which the DnD action was finished.
+        seat: Seat<D>,
+    },
+    /// A client requested to read the server-set selection
+    SendSelection {
+        /// the requested mime type
+        mime_type: String,
+        /// the fd to write into
+        fd: RawFd,
+    },
+}
+
+#[derive(Debug)]
+pub struct DataDeviceUserData {
+    pub(crate) wl_seat: WlSeat,
+}
+
+impl<D> DelegateDispatchBase<WlDataDevice> for DataDeviceState<D> {
```

Comment:
Since we do plan to have server side grabs, it may be best to implement this via ObjectData implementations instead of the dispatch traits (since the high level dispatch machinery only allows one implementation).